File: languages/python/data_structures/stack/stack_list.py
class StackList:
    def __init__(self):
        self._stack = []

    def push(self, item):
        self._stack.append(item)

    def pop(self):
        if self.is_empty():
            raise IndexError("Stack underflow")
        return self._stack.pop()

    def peek(self):
        return self._stack[-1] if not self.is_empty() else None

    def is_empty(self):
        return len(self._stack) == 0

    def size(self):
        return len(self._stack)

File: languages/cpp/data_structures/stack/stack_list.cpp
#include "stack_list.h"

StackList::StackList() : top(nullptr), size(0) {}

StackList::~StackList() {
    while (top != nullptr) {
        Node* temp = top;
        top = top->next;
        delete temp;
    }
}

bool StackList::isEmpty() const {
    return top == nullptr;
}

int StackList::getSize() const {
    return size;
}

void StackList::push(int val) {
    Node* newNode = new Node(val);
    if (top == nullptr) {
        top = newNode;
    } else {
        newNode->next = top;
        top = newNode;
    }
    size++;
}

int StackList::pop() {
    if (top == nullptr) {
        throw std::runtime_error("Trying to pop an empty stack");
    }
    int val = top->data;
    Node* temp = top;
    top = top->next;
    delete temp;
    size--;
    return val;
}

int StackList::peek() const {
    if (top == nullptr) {
        throw std::runtime_error("Trying to peek an empty stack");
    }
    return top->data;
}
